pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,448 @@
+		locations:  make([]*location, len(locs)),
+	}
+
+	for i, loc := range locs {
+		req.locations[i] = &location{
+			address: loc.loc.Address,
+			mapping: &pprof.Mapping{
+				Start:   mapping.MemoryStart,
+				Limit:   mapping.MemoryLimit,
+				Offset:  mapping.FileOffset,
+				BuildID: buildID,
+			},
+		}
+	}
+
+	return req
+}
+
+// updateProfileWithSymbols updates the profile with symbolization results
+func (s *Symbolizer) updateProfileWithSymbols(profile *googlev1.Profile, mapping *googlev1.Mapping, locs []locToSymbolize, symLocs []*location) {
+	stringMap := make(map[string]int64, len(profile.StringTable))
+	for i, str := range profile.StringTable {
+		stringMap[str] = int64(i)
+	}
+
+	type funcKey struct {
+		nameIdx, filenameIdx int64
+	}
+	funcMap := make(map[funcKey]uint64)
+	maxFuncID := uint64(0)
+
+	for _, fn := range profile.Function {
+		if fn.Id > maxFuncID {
+			maxFuncID = fn.Id
+		}
+		funcMap[funcKey{fn.Name, fn.Filename}] = fn.Id
+	}
+
+	for i, symLoc := range symLocs {
+		locIdx := locs[i].idx
+		profile.Location[locIdx].Line = make([]*googlev1.Line, len(symLoc.lines))
+
+		for j, line := range symLoc.lines {
+			nameIdx, ok := stringMap[line.FunctionName]
+			if !ok {
+				nameIdx = int64(len(profile.StringTable))
+				profile.StringTable = append(profile.StringTable, line.FunctionName)
+				stringMap[line.FunctionName] = nameIdx
+			}
+
+			filenameIdx, ok := stringMap[line.FilePath]
+			if !ok {
+				filenameIdx = int64(len(profile.StringTable))
+				profile.StringTable = append(profile.StringTable, line.FilePath)
+				stringMap[line.FilePath] = filenameIdx
+			}
+
+			key := funcKey{nameIdx, filenameIdx}
+			funcID, ok := funcMap[key]
+			if !ok {
+				maxFuncID++
+				funcID = maxFuncID
+				profile.Function = append(profile.Function, &googlev1.Function{
+					Id:        funcID,
+					Name:      nameIdx,
+					Filename:  filenameIdx,
+					StartLine: int64(line.LineNumber),
+				})
+				funcMap[key] = funcID
+			}
+
+			profile.Location[locIdx].Line[j] = &googlev1.Line{
+				FunctionId: funcID,
+				Line:       int64(line.LineNumber),
+			}
+		}
+	}
+
+	mapping.HasFunctions = true
+	mapping.HasFilenames = true
+}
+
+func (s *Symbolizer) symbolize(ctx context.Context, req *request) error {
+	start := time.Now()
+	status := statusSuccess
+	defer func() {
+		s.metrics.profileSymbolization.WithLabelValues(status).Observe(time.Since(start).Seconds())
+	}()
+
+	var table *lidia.Table
+	var err error
+
+	lidiaBytes, err := s.getLidiaBytes(ctx, req.buildID)
+	if err != nil {
+		var bnfErr buildIDNotFoundError
+		if errors.As(err, &bnfErr) {
+			for _, loc := range req.locations {
+				loc.lines = s.createNotFoundSymbols(req.binaryName, loc)
+			}
+			return nil
+		}
+
+		status = statusErrorServerError
+		return err
+	}
+
+	lidiaReader := NewReaderAtCloser(lidiaBytes)
+	table, err = lidia.OpenReader(lidiaReader, lidia.WithCRC())
+	if err != nil {
+		s.metrics.debugSymbolResolution.WithLabelValues("lidia_error").Observe(0)
```

Comment:
I suggest we do not fail the query if there are any errors during symbolization. Let's treat all the errors the same - createNotFoundSymbolz